man/man1/omake.1
.TH omake 1 "15.10.96" "ORL"
.SH NAME
omake \- run GNU make inside an OMNI build tree
.SH SYNOPSIS
.B omake
.I [variable=value]*
.I [target]
.SH DESCRIPTION
The omake command runs GNU make inside an OMNI build tree.  Arguments given to
.B omake
on the command line are passed directly on to
.B make,
but
.B omake
itself generates extra arguments, amongst which are:
.TP 5
.B TOP=<top>
This tells make the path from the current directory to the top of the build
tree.
.TP
.B CURRENT=<current>
This is the path from the top of the tree down to the current directory
.TP
.B VPATH=<directories>
This is a list of the source directories (the equivalent of the current
directory in each of the source trees).  Make will search through these
directories if it can't find a source file in the current directory.
.TP
.B -f <top>/config/config.mk
This tells make to use config.mk as the makefile.  In turn config.mk will
include other files from the import trees and will include dir.mk from either
the current directory or one of the source directories.
.TP
.B -I <source-directory>
These arguments tell make to search the source directories for the dir.mk file.
.TP
.B OMAKE_TARGET=<target>
This is set to the final argument given to omake (usually the target).

.SH SEE ALSO
.BR obuildtree (1)

.SH FILES
.TP 5
.B <top>/config/sources
contains a list of the source trees.
.TP
.B <top>/config/config.mk
is the makefile which
.B omake
uses when building inside this tree.

.SH ENVIRONMENT VARIABLES
.TP 5
.B GNUMAKE
should be set to the command needed to run GNU make if this is different from
.B make

.SH AUTHOR
Tristan Richardson, ORL


ReleaseNotes.txt
RELEASE NOTES
=============

This is omniORB 4.1.0 beta 2.

See README.unix or README.win32.txt for building instructions.
See the documentation in the doc directory for more information.


Changes since omniORB 4.1.0 beta 1
----------------------------------

 - IPv6 support.

     See section 8.6.1 of the omniORB manual for details.

 - New endPointPublish parameter.

     Used for configuring what to publish in IORs. See section 8.6.2
     of the omniORB manual.

 - Minor bug fixes.


Changes since omniORB 4.0.x
---------------------------

 - Objects by value support.

     omniORB now supports the full objects by value specification,
     with the exception of custom valuetypes. See chapter 13 of the
     omniORB manual for information about valuetype in omniORB.

 - Abstract interfaces support.

 - New simpler and more efficient Any implementation.

 - New socket collection implementation.

     Uses poll() on platforms where it is available to avoid limiting
     the number of connections to FD_SETSIZE. Windows specific version
     that better manages the unusual Windows fd_set.

 - New tracing options traceTime and traceInvocationReturns.

     See section 4.2 of the omniORB manual.

 - C++ mapping updated to version 1.1.

     Deprecation of RefCountServantBase; Servant_var template type.

 - Support for local interfaces.

     In particular, ServantManagers and AdapterActivators no longer
     need to be activated in a POA, but can be simple C++ objects. See
     the examples in src/examples/poa/servant_manager.

     Note that to use local interfaces derived from unconstrained
     interfaces, you must compile the idl for the base interfaces
     using the -Wbvirtual_objref -Wbimpl_mapping omniidl flags.



Known issues
------------

This release has the following known issues and omissions:

 - valuetypes supporting local interfaces generate invalid C++. This
   is an issue with the standard C++ mapping.


src/tool/win32/omake.c
/*
 * IMPORTANT - compile this with gnu-win32's gcc, not MS's cl.
 * gcc omake.c -o omake.exe
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>


#define MAX_ARGS 256

void AddToString(char **strp, char *new);


int main(int argc, char **argv)
{
  char top[FILENAME_MAX];
  char current[FILENAME_MAX];
  char currentAbsolute[FILENAME_MAX];
  char tmp[FILENAME_MAX];
  char source[FILENAME_MAX];
  int firstSource = 1;
  char *vpath = NULL;
  char *lastSlash;
  char *makeCmd[MAX_ARGS];
  int nArgs = 0;
  FILE *f;
  int i;
  int quiet = 0;

  if ((argc >= 2) && (strcmp(argv[1],"-quiet") == 0)) {
    quiet = 1;
    argv++;
    argc--;
  }

  getcwd(currentAbsolute,FILENAME_MAX);

  if (strcmp(currentAbsolute,"/") == 0) {
    fprintf(stderr,"ERROR: NOT IN AN OMNI TREE\n");
    exit(1);
  }

  strcpy(top,".");
  strcpy(current,".");

  while (1) {

    if (strcmp(currentAbsolute,"") == 0) {
      fprintf(stderr,"ERROR: NOT IN AN OMNI TREE\n");
      exit(1);
    }

    sprintf(tmp,"%s/config/config.mk",top);

    if (access(tmp, R_OK) == 0) {
      sprintf(tmp,"%s/config/sources",top);

      if (access(tmp, R_OK) == 0) {
	break;
      }
    }

    if (strcmp(top,".") == 0) {

      strcpy(top,"..");

      lastSlash = strrchr(currentAbsolute,'/');
      strcpy(current,lastSlash+1);

      *lastSlash = '\0';	/* strip end off currentAbsolute */

    } else {

      strcpy(tmp,top);
      sprintf(top,"../%s",tmp);

      strcpy(tmp,current);
      lastSlash = strrchr(currentAbsolute,'/');
      sprintf(current,"%s/%s",lastSlash+1,tmp);

      *lastSlash = '\0';	/* strip end off currentAbsolute */
    }
  }

  makeCmd[nArgs++] = "make";
  makeCmd[nArgs++] = "-r";
  makeCmd[nArgs++] = "-f";
  sprintf(tmp,"%s/config/config.mk",top);
  makeCmd[nArgs++] = strdup(tmp);

  sprintf(tmp,"%s/config/sources",top);
  f = fopen(tmp,"r");

  while (fgets(source, FILENAME_MAX, f)) {
    if (source[strlen(source)-1] == '\n') {
      source[strlen(source)-1] = '\0';
    }

    if (source[0] != '/') {	/* source has relative path */
      strcpy(tmp,source);
      sprintf(source,"%s/%s",top,tmp);
    }

    strcpy(tmp,source);
    sprintf(source,"%s/%s",tmp,current);

    if (firstSource) {
      AddToString(&vpath, "VPATH=");
      firstSource = 0;
    } else {
      AddToString(&vpath, ":");
    }

    AddToString(&vpath, source);

    makeCmd[nArgs++] = "-I";
    makeCmd[nArgs++] = strdup(source);
  }

  fclose(f);

  makeCmd[nArgs++] = vpath;

  sprintf(tmp,"TOP=%s",top);
  makeCmd[nArgs++] = strdup(tmp);

  sprintf(tmp,"CURRENT=%s",current);
  makeCmd[nArgs++] = strdup(tmp);

  if (argc > 1) {
    sprintf(tmp,"OMAKE_TARGET=%s",argv[argc-1]);
  } else {
    sprintf(tmp,"OMAKE_TARGET=");
  }
  makeCmd[nArgs++] = strdup(tmp);

  if (!quiet) {
    printf("\n");
    for (i = 0; i < nArgs; i++) {
      printf("%s ",makeCmd[i]);
    }
  }

  for (i = 1; i < argc; i++) {
    makeCmd[nArgs++] = argv[i];
    if (!quiet) {
      printf("'%s' ",argv[i]);
    }
  }

  if (!quiet) {
    printf("\n\n");
  }

  if (execvp(makeCmd[0],makeCmd) < 0) {
    perror("execvp");
    return 1;
  }

  return 0;
}


void AddToString(char **strp, char *new)
{
  if (*strp == NULL) {
    *strp = malloc(strlen(new) + 1);
    *strp[0] = '\0';
  } else {
    *strp = realloc(*strp, strlen(*strp) + strlen(new) + 1);
  }

  strcat(*strp,new);
}
